frontend/class-frontend.php

```diff
@@ -777,33 +858,58 @@ function robots_for_single_post( $robots, $postid = 0 ) {
 		 * @return string $canonical
 		 */
 		public function canonical( $echo = true, $un_paged = false, $no_override = false ) {
-			$canonical       = false;
-			$skip_pagination = false;
+			if ( is_null( $this->canonical ) ) {
+				$this->generate_canonical();
+			}
+
+			$canonical = $this->canonical;
+
+			if ( $un_paged ) {
+				$canonical = $this->canonical_unpaged;
+			}
+
+			if ( $no_override ) {
+				$canonical = $this->canonical_no_override;
+			}
+
+			if ( $echo !== false ) {
+				echo '<link rel="canonical" href="' . esc_url( $canonical, null, 'other' ) . '" />' . "\n";
+			} else {
+				return $canonical;
+			}
+		}
+
+		/**
+		 * This function normally outputs the canonical but is also used in other places to retrieve
+		 * the canonical URL for the current page.
+		 *
+		 * @return string $canonical
+		 */
+		private function generate_canonical() {
+			$canonical          = false;
+			$canonical_override = false;
 
 			// Set decent canonicals for homepage, singulars and taxonomy pages
 			if ( is_singular() ) {
-				$meta_canon = WPSEO_Meta::get_value( 'canonical' );
-				if ( $no_override === false && $meta_canon !== '' ) {
-					$canonical       = $meta_canon;
-					$skip_pagination = true;
-				} else {
-					$obj       = get_queried_object();
-					$canonical = get_permalink( $obj->ID );
-
-					// Fix paginated pages canonical, but only if the page is truly paginated.
-					if ( get_query_var( 'page' ) > 1 ) {
-						global $wp_rewrite;
-						$numpages = substr_count( $obj->post_content, '<!--nextpage-->' ) + 1;
-						if ( $numpages && get_query_var( 'page' ) <= $numpages ) {
-							if ( ! $wp_rewrite->using_permalinks() ) {
-								$canonical = add_query_arg( 'page', get_query_var( 'page' ), $canonical );
-							} else {
-								$canonical = user_trailingslashit( trailingslashit( $canonical ) . get_query_var( 'page' ) );
-							}
+				$obj       = get_queried_object();
+				$canonical = get_permalink( $obj->ID );
+
+				$this->canonical_unpaged = $canonical;
+
+				$canonical_override = WPSEO_Meta::get_value( 'canonical' );
+
+				// Fix paginated pages canonical, but only if the page is truly paginated.
+				if ( get_query_var( 'page' ) > 1 ) {
+					global $wp_rewrite;
```

Comment:
Maybe put this under line 905, because the variable is used in that if-statement.